### ISSUE TITLE
The paddings in multihead attention model

### ISSUE THREAD

**Issue by lsyysl9711** [dmlc/dgl#2657]
## ❓ Questions and Help

When I use set_transformer model on graphs implemented by DGL, the model will output NaN just after one training epoch. After reading source codes for a few hours, I think I find where the mistakes happen.

 # padding to (B, max_len_x/mem, num_heads, d_head)
        queries = F.pad_packed_tensor(queries, lengths_x, 0)
        keys = F.pad_packed_tensor(keys, lengths_mem, 0)
        values = F.pad_packed_tensor(values, lengths_mem, 0)

# attention score with shape (B, num_heads, max_len_x, max_len_mem)
        e = th.einsum('bxhd,byhd->bhxy', queries, keys)
        # normalize
        e = e / np.sqrt(self.d_head)

        # generate mask
        mask = th.zeros(batch_size, max_len_x, max_len_mem).to(e.device)
        for i in range(batch_size):
            mask[i, :lengths_x[i], :lengths_mem[i]].fill_(1)
        mask = mask.unsqueeze(1)
        e.masked_fill_(mask == 0, -float('inf'))

The above are the codes that I think finally cause this problem. Since we need to pad neighbors of every node so that they have the same number of neighbors, there will be multiple rows consist of zeros only. And then we will use "-inf" to mask Q,K,V. So the rows consist of only zeros will consist of "-inf" only. When these rows are sent to softmax function later, they will generate many rows consist of "NaN". And we need to multiply several matrix, so after several epochs we will only have matrix consists of NaN. So how can I solve this problem, using a different number to pad the sequence?

**Comment by yzh119** [comment]
@lsyysl9711 thanks for reporting this bug, yes I do think you are right and I'm working on fixing this bug.

**Comment by yzh119** [comment]
One line of `alpha[alpha.isnan()]= 0.` should work IMO.

**Comment by lsyysl9711** [comment]
Thank you so much for replying me. I tried your method and the compiler gave me the error information "Tensor object has no attribute isnan". By the way, I roughly think we need to take some more complex cases into consideration: Given a matrix A, what we need to do is to prevent the rows that only contain NaN from calculating softmax. For other rows that both contain NaN and normal numeric values, we should not remove the NaN in them. So I think "alpha[alpha.isnan()]" might remove all of the NaN in the matrix.(Maybe I misunderstand your solutions, if I do not get your point, feel free to leave a message).

**Comment by yzh119** [comment]
@lsyysl9711 yes your concern is right and I think `alpha.masked_fill_(mask == 0, 0.)` is what we need.

btw, `isnan()` should be available in PyTorch 1.7, what version of PyTorch you were using?

**Comment by lsyysl9711** [comment]
Hmm, I should update my pytorch version. I think this would work, but when I run the codes the compiler gave me error information: one of the variables needed for gradient computation has been modified by an inplace operation [torch.FloatTensor [2708, 4, 1, 1]], which is output 0 of SoftmaxBackward, is at version 1; expected version 0 instead. So I think we may need more modification. >.<

**Comment by lsyysl9711** [comment]
I think this compiler error may be caused by rewriting outputs of softmax function.

**Comment by lsyysl9711** [comment]
Maybe we should use non in-place operation such as torch.where ? If you have other ideas, please feel free to leave a message.

**Comment by yzh119** [comment]
`masked_fill_` is a in-place operation, try replacing it with `alpha = alpha.masked_fill(mask == 0, 0.)`, see the changes in #2658.

**Comment by lsyysl9711** [comment]
Thanks
